src/container/cli.c

```diff
@@ -973,6 +973,124 @@ dc_cont_query(tse_task_t *task)
 	return rc;
 }
 
+struct cont_set_prop_args {
+	struct dc_pool		*cqa_pool;
+	struct dc_cont		*cqa_cont;
+	crt_rpc_t		*rpc;
+	daos_handle_t		hdl;
+};
+
+static int
+cont_set_prop_complete(tse_task_t *task, void *data)
+{
+	struct cont_set_prop_args	*arg = (struct cont_set_prop_args *)
+						data;
+	struct cont_prop_set_out	*out = crt_reply_get(arg->rpc);
+	struct dc_pool			*pool = arg->cqa_pool;
+	struct dc_cont			*cont = arg->cqa_cont;
+	int				 rc   = task->dt_result;
+
+	rc = cont_rsvc_client_complete_rpc(pool, &arg->rpc->cr_ep, rc,
+					   &out->cpso_op, task);
+	if (rc < 0)
+		D_GOTO(out, rc);
+	else if (rc == RSVC_CLIENT_RECHOOSE)
+		D_GOTO(out, rc = 0);
+
+	if (rc != 0) {
+		D_ERROR("RPC error while setting prop on container: "DF_RC"\n",
+			DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+
+	rc = out->cpso_op.co_rc;
+
+	if (rc != 0) {
+		D_DEBUG(DF_DSMC, DF_CONT": failed to set prop on container: "
+			"%d\n",
+			DP_CONT(pool->dp_pool, cont->dc_uuid), rc);
+		D_GOTO(out, rc);
+	}
+
+	D_DEBUG(DF_DSMC, DF_CONT": Set prop: using hdl="DF_UUID"\n",
+		DP_CONT(pool->dp_pool, cont->dc_uuid),
+		DP_UUID(cont->dc_cont_hdl));
+
+out:
+	crt_req_decref(arg->rpc);
+	dc_cont_put(cont);
+	dc_pool_put(pool);
+	return rc;
+}
+
+int
+dc_cont_set_prop(tse_task_t *task)
+{
+	daos_cont_set_prop_t		*args;
+	struct cont_prop_set_in		*in;
+	struct dc_pool			*pool;
+	struct dc_cont			*cont;
+	crt_endpoint_t			 ep;
+	crt_rpc_t			*rpc;
+	struct cont_set_prop_args	 arg;
+	int				 rc;
+
+	args = dc_task_get_args(task);
+	D_ASSERTF(args != NULL, "Task Argument OPC does not match DC OPC\n");
+
+	cont = dc_hdl2cont(args->coh);
+	if (cont == NULL)
+		D_GOTO(err, rc = -DER_NO_HDL);
+
```

Comment:
can cont->dc_capas be checked to see if it includes DAOS_COO_RW in this function, and return an error in this event that the container was not opened with RW permissions?

Reply:
We could check here, but we must gate access on the server side also, as in the current implementation -- need to cover for the case where a client uses their own modified version of libdaos that takes out the check. Is there a reason to kick out the user early? I'm reluctant to check the same condition twice.

Reply:
Only reason I can think of is to quickly error directly at the client, avoid the RPC trip to the pool service leader server? Not sure it's a big deal. I agree with you the check already in the server side needs to remain there.

I had been looking at some other client-side code I think is must have been cont_open(). it checks if the container open argument flag is RW meanwhile if the pool capas is RO then the client fails directly with -DER_NO_PERM before sending RPC.

Reply:
Per our Skype chat, leaving it as-is for now. We can address the client side when we begin enforcing the container ACLs, which will require changes to this check.